{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5e76d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from dateutil.parser import parse\n",
    "from collections import Counter\n",
    "from dateutil import parser\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "876d4aa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "930"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#downloaded older sequences (only up to 2016) including shorter sequences\n",
    "#will make gene-specific builds. \n",
    "#Need to find all accessions that I don't already have and curate them into a new data file\n",
    "\n",
    "#sequence data downloaded as a multi-genbank file\n",
    "#compile all sequences into a fasta file with header: accession|strain_name|date|country|host|subtype\n",
    "#Put date in 'yyyy-mm-dd' format\n",
    "\n",
    "\n",
    "#find all full genome sequences that are already in the data file            \n",
    "full_genome_records = []\n",
    "full_genome_accesions = []\n",
    "\n",
    "for record in SeqIO.parse(open(f\"rhinovirusC_all.fasta\",\"r\"), \"fasta\"):\n",
    "    accession = record.description.split('|')[0]\n",
    "    full_genome_accesions.append(accession)\n",
    "    full_genome_records.append(SeqRecord(record.seq, id=record.description, description=''))\n",
    "\n",
    "\n",
    "not_subtyped_count = 0\n",
    "subtyped_count = 0\n",
    "\n",
    "total=0\n",
    "new=0\n",
    "\n",
    "\n",
    "#store all edited sequence records\n",
    "seq_records = []\n",
    "for record in SeqIO.parse(open(f\"genbank_older_seqs_partialcds.gb\",\"r\"), \"genbank\"):\n",
    "    total+=1\n",
    "    accession = record.annotations['accessions'][0]\n",
    "    #if I don't already have this sequence\n",
    "    if accession not in full_genome_accesions:\n",
    "        new+=1\n",
    "        collection_date, strain_name, country, host, subtype = 'None', 'None', 'None', 'None', 'None'\n",
    "        #exclude patent and synthetic sequences, that are not clinical isolates\n",
    "        if record.annotations['data_file_division'] not in ['PAT', 'SYN']:\n",
    "\n",
    "            for feature in record.features:\n",
    "                if 'collection_date' in feature.qualifiers:\n",
    "                    collection_date = feature.qualifiers['collection_date'][0]   \n",
    "                if 'strain' in feature.qualifiers:\n",
    "                    strain_name = feature.qualifiers['strain'][0]\n",
    "                if 'country' in feature.qualifiers:\n",
    "                    country = feature.qualifiers['country'][0]\n",
    "                if 'host' in feature.qualifiers:\n",
    "                    host = feature.qualifiers['host'][0]\n",
    "\n",
    "            #find subtype from 'organism' genbank entry\n",
    "            subtype_organism_search = re.search('rhinovirus C([0-9]*)', record.annotations['organism'], re.IGNORECASE)\n",
    "\n",
    "            if subtype_organism_search:\n",
    "                subtype = subtype_organism_search.group(1)\n",
    "                if subtype == '':\n",
    "                    subtype = 'None'\n",
    "\n",
    "            #otherwise, look in the 'description'\n",
    "            if subtype == 'None':\n",
    "                subtype_description_search = re.search('rhinovirus C strain C([0-9]*)', record.description, re.IGNORECASE)\n",
    "\n",
    "                if subtype_description_search:\n",
    "                    subtype = subtype_description_search.group(1)\n",
    "                    if subtype == '':\n",
    "                        subtype = 'None'\n",
    "\n",
    "            if subtype == 'None':\n",
    "                not_subtyped_count+=1\n",
    "            else:\n",
    "                subtyped_count+=1\n",
    "\n",
    "\n",
    "\n",
    "            #only keep country information, not city/state/region\n",
    "            if ':' in country:\n",
    "                country = country.split(':')[0]\n",
    "\n",
    "            if collection_date == 'None':\n",
    "                collection_date = record.annotations['date']\n",
    "\n",
    "\n",
    "            #only keep sequences with date\n",
    "            if collection_date != 'None' and collection_date!='May-2016/Dec-2017':\n",
    "                formatted_date = parser.parse(collection_date).strftime('%Y-%m-%d')\n",
    "                #dateutil parser will assign a day (today's date) to unknown days, and same for month, want XX instead\n",
    "                if len(collection_date)==8:\n",
    "                    formatted_date = formatted_date[:-2] + 'XX'\n",
    "                elif len(collection_date)==4:\n",
    "                    formatted_date = formatted_date[:5] + 'XX-XX'\n",
    "\n",
    "                list_of_info = [accession, strain_name, formatted_date, country, host, subtype]\n",
    "                new_record_info = '|'.join(list_of_info)\n",
    "                seq_records.append(SeqRecord(record.seq, id=new_record_info, description=''))  \n",
    "\n",
    "all_seq_records = full_genome_records + seq_records\n",
    "\n",
    "#add new (partial genome) sequences to those I already have (full genome)\n",
    "#write fasta sequence file \n",
    "SeqIO.write(all_seq_records, f'rhinovirusC_partialcds_all.fasta', \"fasta\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
